doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+
+### Architecture Design
+
+No architecture change is needed for Master Arbitration. The protobuf changes is done already. The protobuf of a `MasterArbitration` message is defined as following:
+
+```
+message MasterArbitration {
+  Role role = 1;
+  Uint128 election_id = 2;
+}
+
+message Uint128 {
+  uint64 high = 1;
+  uint64 low = 2;
+}
+
+message Role {
+  string id = 1;
+}
+```
+
+The Master Arbitration is an extension of the basic gNMI protocol. It uses the extension field in the SetRequest message that has been introduced to allow adding such extensions as the Master Arbitration.
+
+```
+// SetRequest is sent from a client to the target to update values in the data
+// tree. Paths are either deleted by the client, or modified by means of being
+// updated, or replaced. Where a replace is used, unspecified values are
+// considered to be replaced, whereas when update is used the changes are
+// considered to be incremental. The set of changes that are specified within
+// a single SetRequest is considered to be a transaction.
+// Reference: gNMI Specification Section 3.4.1
+message SetRequest {
+  ...
+  // Extension messages associated with the SetRequest. See the
+  // gNMI extension specification for further definition.
+  repeated gnmi_ext.Extension extension = 5;
+}
+```
+
+The Master Arbitration uses the following messages that are defined in the `gnmi_ext.proto`.
+
+```
+// ...
+
+// The Extension message contains a single gNMI extension.
+message Extension {
+  oneof ext {
+    RegisteredExtension registered_ext = 1;    // A registered extension.
+    // Well known extensions.
+    MasterArbitration master_arbitration = 2;  // Master arbitration extension.
+    History history = 3;                       // History extension.
+  }
+}
+
+// ...
+```
+
+### High-Level Design
+
+When Master Arbitration is enabled, `SetRequest` with `MasterArbitration` message sent from clients (controllers) will get checked by gNMI. The client with the highest `election_id` will be treated as master with the condition of `ReqFromMaster = True`. The rest clients are non-masters. After the initial `SetRequest`, only request from master will be handled by gNMI server, the requests from non-master clients will be dropped.
+
+The implementation of Master Arbitration, a new built-in feature in SONIC, will involve modifications in the telemetry module and the sonic-gnmi repository. However, it is important to note that this change will not affect other modules or sub-modules. It is worth mentioning that Master Arbitration does not require any changes to be made to SWSS or Syncd. A new field will be created in `CONFIG_DB` to record the status of Master Arbitration.
```

Comment:
Is this new field role-scoped?

Reply:
This new field is recording the status of Master Arbitration, say {"with-master-arbitration": true} or {"with-master-arbitration": false}. No plan to make it role-scoped.

Reply:
Is there any possibility of receiving set requests from clients with different roles and EIDs to the same switch? if yes, do we need to store all these info. into the config_DB.

Reply:
Currently we only have one role, which is the default one. For the detail about multiple-role case, we would like to include in the future release.